Zend/zend_vm_def.h

```diff
@@ -4024,6 +4016,17 @@ ZEND_VM_HOT_HANDLER(131, ZEND_DO_FCALL_BY_NAME, ANY, ANY, SPEC(RETVAL))
 				zend_verify_internal_return_type(call->func, ret));
 			ZEND_ASSERT((call->func->common.fn_flags & ZEND_ACC_RETURN_REFERENCE)
 				? Z_ISREF_P(ret) : !Z_ISREF_P(ret));
+
+			/* Type checks for internal functions are usually only performed by zpp.
+			 * In debug mode we additionally run arginfo checks to detect cases where
+			 * arginfo and zpp went out of sync. */
+			if (UNEXPECTED(fbc->common.fn_flags & ZEND_ACC_HAS_TYPE_HINTS)
+			 && UNEXPECTED(!zend_verify_internal_arg_types(fbc, call))) {
+				zend_vm_stack_free_call_frame(call);
+				zend_rethrow_exception(execute_data);
+				UNDEF_RESULT();
+				HANDLE_EXCEPTION();
+			}
```

Comment:
This movement of argument checks after the actual call, looks very suspicious.
Internal functions may convert arguments in place.